### PR TITLE
Harmony 388 - Better consistency across Python repos

### PR DESCRIPTION
**Pull request by vinnyinverso**
What follows are the acceptance criteria and what I found or did to address them. (I also spoke with Chris prior to submitting this PR.)
- Consistent python environment setup and use across the 3 repos 
  - For all but the harmony-service example, the READMEs were suggesting to use some combination of **pyenv**, **pyenv-virtualenv**, or **venv**. I tried to update the READMEs so that it was clear to the reader that all of these were acceptable tools.
- Either build **scripts** or **make** across the 3 repos
  - All of the regular (non-Docker service) Python module/lib repos (harmony-service-lib, harmony-py) use make files. All of the Docker/service-oriented repos (harmony-netcdf-to-zarr, harmony-service-example) use build scripts because they seem more conducive to the scripting logic needed to satisfy some of the more complex build logic. This seemed like a reasonably consistent solution (makefiles for pure libraries and build scripts for docker services).
- Consistent way of installing deps across the 3 repos
  -  All except one repo are using **pip** for installing dependencies in a fairly consistent manner, with one requirements file for dev, and one for core requirements. The only repo not following this pattern is the harmony-service-example, which uses **conda**. However, conda is a recommended way of installing one of the main dependencies of this repo (gdal) whereas pip is not (and is difficult to use in this instance from my experience). The use of conda, is also still compatible with the aforementioned recommended Python version manager and virtual env tools.
- Consistent linting across the 3 repos
  - All repos (except for harmony-service-example) use **flake8**. We could add **flake8** to harmony-service-example if desirable. I'm guessing it wasn't added due to the simplicity of the example.
- Consistent unit testing framework and execution across the 3 repos
  - I believe Patrick said there was another ticket addressing this.
- Consistent CI/CD across the 3 repos
  - Much of this seems to be adequately addressed by the recent transition to **GitHub** and related work.
- Update READMEs across the 3 repos to document the consistent tools, packages, and configuration for setting up a dev environment; dev environment does not presume editor
  -  I updated 3 READMEs to give more consistent suggestions on virtual environment setup. None of the dev environment setups assume the user has a specific editor.

### CODE REVIEW

**Review by chris-durbin — APPROVED** [review]

**Review by bilts — CHANGES_REQUESTED** [review]
Individually, these PRs are good.  I think they're missing some things when taken together, though.

A couple of remaining inconsistencies that I think should be addressed:

1. How we're building things in the build scripts.  They're fundamentally doing the same thing, but they're doing it two completely different ways.  A consequence of this came up recently when our service example ended up with an outdated version of the service library because we didn't notice its build was working differently than netcdf-to-zarr.
2. (Related) netcdf-to-zarr uses a separate Dockerfile for building with a local service lib, while harmony-service-example includes a local directory
2. Test suite (or lack thereof) aside, the location of the test folder is different in the two projects.  harmony-service-example should have its tests pulled to the top level folder like our other projects.
3. harmony-netcdf-to-zarr and harmony-service-lib-py express their dependencies in .txt files for use with pip.  harmony-service-example uses a yml file for use with conda.  harmony-py uses a pyproject.toml file for use with poetry.
4. While the projects generally use .flake8 for linting, their settings (.flake8 files) are different
5. `harmony-py` has a `Makefile`, other projects have a `makefile`
6. The makefiles unnecessarily differ in their available tasks.  `harmony-py` has `make install` whereas the equivalent in `harmony-service-lib-py` is `make develop`
7. The two Python libraries have different manifests
8. The .gitignore files are mostly similar boileplate, but some have customizations at the top, some at the bottom, and (IIRC) some may have some wherever they seem to belong in the middle.

To sum it all up, I don't think we need to change code, but there's a lot of boilerplate that's differing in ways that make it hard to get bearings.

**Comment by indiejames** [comment]
I started peeling off Patricks list starting with the easiest stuff. 
1. Moved 'tests' directory to top level for 'harmony-service-example'.
2. Added flake8 linter config and fixed resulting linter errors

I'm pushing these changes to Vinny's HARMONY-388 branches on the respective repos.

**Comment by vinnyinverso** [comment]
Just keeping track here -- thanks for the help @indiejames 

- [x] How we're building things in the build scripts.  They're fundamentally doing the same thing, but they're doing it two completely different ways.  A consequence of this came up recently when our service example ended up with an outdated version of the service library because we didn't notice its build was working differently than netcdf-to-zarr.
- [x] (Related) netcdf-to-zarr uses a separate Dockerfile for building with a local service lib, while harmony-service-example includes a local directory
- [x] Test suite (or lack thereof) aside, the location of the test folder is different in the two projects.  harmony-service-example should have its tests pulled to the top level folder like our other projects.
- [x] harmony-netcdf-to-zarr and harmony-service-lib-py express their dependencies in .txt files for use with pip.  harmony-service-example uses a yml file for use with conda.  harmony-py uses a pyproject.toml file for use with poetry.
- [x] While the projects generally use .flake8 for linting, their settings (.flake8 files) are different
- [x] `harmony-py` has a `Makefile`, other projects have a `makefile`
- [x] The makefiles unnecessarily differ in their available tasks.  `harmony-py` has `make install` whereas the equivalent in `harmony-service-lib-py` is `make develop`
- [x] The two Python libraries have different manifests
- [x] The .gitignore files are mostly similar boileplate, but some have customizations at the top, some at the bottom, and (IIRC) some may have some wherever they seem to belong in the middle.

**Comment by vinnyinverso** [comment]
@bilts @chris-durbin I made some more changes taking into account the feedback from here and the slack group chat. Here are the other PRs:

https://github.com/nasa/harmony-py/pull/34
https://github.com/nasa/harmony-netcdf-to-zarr/pull/5
https://github.com/nasa/harmony-service-lib-py/pull/13

I also created a short/consolidated [document on venv](https://github.com/nasa/harmony-service-example/blob/HARMONY-388/ENVHELP.md) guidance (pulling together the guidance I found across repos in a consistent manner) so that we can simplify the READMEs/setup. Rather than repeating a bunch of verbiage on virtual environments 4 times, the READMEs would point to this single doc now. It's probably more of a cheatsheet than anything.

The only packaging tools used now are pip and in one instance conda (because gdal installation is difficult otherwise). That being said, in both cases the setup/installation of dependencies is nearly identical.

Feel free to reach out with any questions or suggestions. (I still want to test things a bit more locally to make sure the Makefiles work as expected, once I switch back to Harmony work.)

File: README.md [hunk 1]
```diff
@@ -52,25 +43,25 @@ Then install it into your conda environment in development mode. Subsequent chan
...
 1. Build the Docker image (installs Harmony Service Library from PyPI):
 
-        $ bin/build-image
+        $ make build-image
 
 If the Docker build does not complete and or this breaks your local Docker
 environment, try increasing the memory allocated to your Docker environment.
 
-If you'd like the Docker image to include a local version of the Harmony Service Library, set the `LOCAL_SERVICE_LIB` environment variable to its location and build:
+If you'd like the Docker image to include a local version of the Harmony Service Library, set the `LOCAL_SVCLIB_DIR` environment variable to its location and build:
```
Comment (chris-durbin) [inline README.md:52]: FYI - you'll need to update all of the Bamboo deployments for the harmony-service-example to set this variable prior to deployment:

They all currently execute:
```
LOCAL_SERVICE_LIB=./harmony-service-lib-py bin/build-image
```

**Review by chris-durbin — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (vinnyinverso) [reply README.md:52]: Updated to `LOCAL_SERVICE_LIB=./harmony-service-lib-py LOCAL_SVCLIB_DIR=./harmony-service-lib-py bin/build-image` in bamboo scripts as you suggested. Will clean this up after everything is deployed.

Comment (bilts) [reply README.md:52]: Good catch.  After merge, I'm anticipating wanting/needing to do the same activity in our Bamboo builds and deployments. It seems likely something will break and they at least need a look.  Our services should be running the same steps as one another, as should our Python libs

**Review by bilts — APPROVED** [review]

**Pull request merged**